metacoadd/utils.py

```diff
@@ -1,25 +1,94 @@
-import copy
+from typing import Optional
 
+import astropy.wcs
 import galsim
 import ngmix
 import numpy as np
 
 
-def shift_wcs(wcs, offset):
-    # TODO: check inputs
+class WCSBundle:
+    _astropy_wcs: astropy.wcs.WCS
+    _galsim_WCS: galsim.BaseWCS
+
+    def __init__(
+        self,
+        wcs: astropy.wcs.WCS | galsim.BaseWCS,
+        image: Optional[np.ndarray | galsim.Image] = None,
+    ):
+        """Create a bundled wcs containing both astropy and galsim wcs.
+
+        The provided wcs can be either an astropy wcs or a galsim wcs.
+        If wcs is an astropy wcs: create a galsim.AstropyWCS (which is a
+        BaseWCS) and keep both.
+
+        If wcs is a galsim wcs:
+            - image must be given and be either an ndarray or a galsim Image
+            - the astropy wcs will be created from the galsim wcs and the image
+              bounds (either using image.bounds if galsim image or
+              BoundsI(1, 1 , nrow, ncol) if ndarray).
+
+        The wcs can be accessed with properties wcs_bundle.astropy or
+        wcs_bundle.galsim
+        """
+        if isinstance(wcs, astropy.wcs.WCS):
+            self._astropy_wcs = wcs
+            self._galsim_wcs = galsim.AstropyWCS(header=wcs.to_header())
+        elif isinstance(wcs, galsim.BaseWCS):
+            self._galsim_wcs = wcs
+            # in this case, we need the image
+            if isinstance(image, np.ndarray):
+                # we have no wcs in this image
+                xmin, ymin = 1, 1
+                nrow, ncol = image.shape
+                bounds = galsim.BoundsI(
+                    xmin, xmin + ncol - 1, ymin, ymin + nrow - 1
+                )
+                header = {}
+                wcs.writeToFitsHeader(header, bounds)
+                self._astropy_wcs = astropy.wcs.WCS(header)
+            elif isinstance(image, galsim.Image):
+                # There may be a wcs in this image, check they are compatible
+                if image.wcs is not None:
+                    reset_wcs = False
+                    if image.wcs != wcs:
+                        raise ValueError(
+                            "image wcs and provided wcs are different"
+                        )
+                else:
+                    reset_wcs = True
+                    image.wcs = wcs
+                header = {}
+                image.wcs.writeToFitsHeader(header, image.bounds)
+                self._astropy_wcs = astropy.wcs.WCS(header)
+                if reset_wcs:
+                    image.wcs = None
+            else:
+                raise ValueError(
+                    "missing image or wrong type, needed when giving galsim wcs"
```

Comment:
Logic could be a bit simplified here.

Also I'm wondering if `reset_wcs` is really necessary.

Reply:
I was hesitating too about reset_wcs, but I don't like a function to modify its argument if it is not its goal (side effects are bad)... The logic was complex (I just used the existing one which was spreader among different functions), and it looks a bit fragile to me. And the only reason for it is to keep the astropy.wcs.WCS for use in `reproject` module... Maybe it can be simplified